### PR TITLE
extras can update itself with this

### PR DESCRIPTION
**Pull request by michaelneale**
@markhibberd @nhibberd a self upgrade-ey thing (using one weird old trick for fetching latest from github).

### CODE REVIEW

File: src/cabal [hunk 1]
```diff
@@ -108,6 +109,23 @@ run_repl () {
...
+# Upgrade the script in place
+#
+run_upgrade () {
+    echo "Checking for a new version of cabal extras ..."
+    TZ=$(date +"%T")
+    curl --silent "https://raw.githubusercontent.com/ambiata/cabal-extras/master/src/cabal?$TZ" > /tmp/cabal
+    if ! cmp ./cabal /tmp/cabal  >/dev/null 2>&1
+    then
+      echo "New version found and upgraded. You can now commit it to your git repo."
+      mv /tmp/cabal ./cabal
```
Comment (charleso) [inline src/cabal]: One thing we've wanted to do for a while is insert a hash comment with the version in github. /feature-request

Comment (michaelneale) [inline src/cabal]: yeah mark mentioned that - any ideas how?

Comment (markhibberd) [inline src/cabal]: This one is probably well off the hack-ometer, but only way I can think of is:
```
curl https://api.github.com/repos/ambiata/cabal-extras/commits/HEAD | parse some json with sed
```

Comment (markhibberd) [inline src/cabal]: I can't believe I qualified with *probably*.

Comment (michaelneale) [inline src/cabal]: good god

Comment (charleso) [inline src/cabal]:
```
git ls-remote  https://github.com/ambiata/cabal-extras | grep refs/heads/master | cut -f 1
```

Comment (domdere) [inline src/cabal]: be nice if we could also keep `Setup.hs` up to date too

Comment (michaelneale) [inline src/cabal]: @domdere yeah similar approach could be used

Comment (michaelneale) [inline src/cabal]: @charleso yeah that works - but not sure how to shoehorn the commit sha into the script itself (given it won't be known until merged into cabal-extras)

**Comment by michaelneale** [comment]
@charleso @markhibberd hows this sick bash for comparing (ignoring version placeholder) and then inserting a commit number? (so only the copied cabal script in the client project will have the commit number from the extras repo)

**Comment by tmcgilchrist** [comment]
This looks great

File: src/cabal [hunk 2]
```diff
@@ -1,3 +1,4 @@
+# EXTRAS_VERSION
```
Comment (charleso) [inline src/cabal]: Dumb question - there's not problem if the first line doesn't contain the `#!`?

File: src/cabal [hunk 3]
```diff
@@ -108,6 +110,24 @@ run_repl () {
 }
 
 #
+# Upgrade the script in place
+#
+run_upgrade () {
+    echo "Checking for a new version of cabal extras ..."
+    TZ=$(date +"%T")
+    curl --silent "https://raw.githubusercontent.com/ambiata/cabal-extras/master/src/cabal?$TZ" > /tmp/cabal
+    if ! cmp <(tail -n +2 ./cabal) <(tail -n +2 /tmp/cabal)  >/dev/null 2>&1
```
Comment (charleso) [inline src/cabal]: :pager:

File: src/cabal (hunk 2 above)
Comment (michaelneale) [inline src/cabal]: hrm yeah - well it may cause some issues... I will look at another way.

Comment (michaelneale) [inline src/cabal]: yes, <(...) is bash only, and moving the #! down means that it defaults to bash (at least on OSX), hence my hack accidentally worked. Will do it differently. Lucky you asked and I checked.

**Comment by michaelneale** [comment]
@charleso ok I have it appending the version string as suggested. And no damage done by accidentally making it a bash script. Look ok?

**Comment by charleso** [comment]
:+1:

**Comment by michaelneale** [comment]
I'll upgrade the template projects with this tomorrow before too many new projects get created. Then upgrading will be easy.

**Pull request merged**